src/Simulation/QSharpFoundation/Diagnostics/Assert.qs

```diff
@@ -86,6 +86,7 @@ namespace Microsoft.Quantum.Diagnostics {
     operation AssertMeasurementProbability(bases : Pauli[], qubits : Qubit[], result : Result, prob : Double, msg : String, tolerance : Double) : Unit
     is Adj + Ctl {
         body intrinsic;
+        adjoint self;
     }
```

Comment:
Doesn't fragment `adjoint self` duplicate the fragment `is Adj`?

Reply:
Actually, the default directive for the adjoint specialization is `adjoint invert` which will reverse the order of instructions in an operation and uses their adjoint. So if you have an operation defined like this:
```qsharp
operation Foo(target : Qubit) : Unit is Adj {
    A(target);
    B(target);
}
```
That is the equivalent of:
```qsharp
operation Foo(target : Qubit) : Unit is Adj {
    body (...) { 
        A(target);
        B(target);
    }
    adjoint (...) {
        Adjoint B(target);
        Adjoint A(target);
    }
}
```
Whereas `adjoint self` indicates that the body specialization can be used as the adjoint and allows the Q# compiler to skip generating an adjoint at all. So it effectively makes it such that:
```qsharp
operation Foo(target : Qubit) : Unit is Adj {
    body (...) {
        A(target);
        B(target);
    }
    adjoint self;
}
```
is equivalent to:
```qsharp
operation Foo(target : Qubit) : Unit is Adj {
    body (...) {
        A(target);
        B(target);
    }
    adjoint (...) {
        Foo(target);
    }
}
```

For this specific case, declaring `AssertMeasurementProbability` as `adjoint self` means that no adjoint specialization will be generated, each instance of `Adjoint AssertMeasurementProbability` will just forward to the body specialization, and we don't have to implement either the adjoint or controlled adjoint qis functions for it.